File: implementations/python/src/foilbench_py/core/geometry.py
"""NACA four-digit geometry and moving-boundary queries."""

from dataclasses import dataclass

import numpy as np
from jaxtyping import Bool, Float

from foilbench_py.core.models import ControlState, DomainSpec, FoilSpec
from foilbench_py.types import PointCloud


@dataclass(frozen=True, slots=True)
class NacaFoil:
    spec: FoilSpec

    @property
    def maximum_camber(self) -> float:
        return int(self.spec.naca[0]) / 100.0

    @property
    def camber_position(self) -> float:
        return int(self.spec.naca[1]) / 10.0

    @property
    def thickness(self) -> float:
        return int(self.spec.naca[2:]) / 100.0

    @property
    def maximum_radius(self) -> float:
        vertical_extent = (
            self.maximum_camber + 0.51 * self.thickness
        ) * self.spec.chord
        return float(np.hypot(0.75 * self.spec.chord, vertical_extent))

    def _to_local(self, points: PointCloud, angle_degrees: float) -> Float[np.ndarray, "point 2"]:
        angle = np.deg2rad(angle_degrees)
        cosine = np.cos(angle)
        sine = np.sin(angle)
        translated = points[:, :2] - np.asarray(self.spec.pivot[:2], dtype=points.dtype)
        local = np.empty_like(translated)
        local[:, 0] = cosine * translated[:, 0] + sine * translated[:, 1]
        local[:, 1] = -sine * translated[:, 0] + cosine * translated[:, 1]
        local[:, 0] += 0.25 * self.spec.chord
        return local

    def _from_local(
        self, points: Float[np.ndarray, "point 2"], angle_degrees: float
    ) -> Float[np.ndarray, "point 2"]:
        angle = np.deg2rad(angle_degrees)
        cosine = np.cos(angle)
        sine = np.sin(angle)
        shifted = points.copy()
        shifted[:, 0] -= 0.25 * self.spec.chord
        world = np.empty_like(shifted)
        world[:, 0] = cosine * shifted[:, 0] - sine * shifted[:, 1]
        world[:, 1] = sine * shifted[:, 0] + cosine * shifted[:, 1]
        world += np.asarray(self.spec.pivot[:2], dtype=points.dtype)
        return world

    def surfaces(
        self, x_local: Float[np.ndarray, " point"]
    ) -> tuple[
        Float[np.ndarray, " point"],
        Float[np.ndarray, " point"],
    ]:
        chord = self.spec.chord
        x = np.clip(x_local / chord, 0.0, 1.0)
        thickness = (
            5.0
            * self.thickness
            * chord
            * (
                0.2969 * np.sqrt(np.maximum(x, 0.0))
                - 0.1260 * x
                - 0.3516 * x**2
                + 0.2843 * x**3
                - 0.1036 * x**4
            )
        )
        m = self.maximum_camber
        p = self.camber_position
        camber = np.zeros_like(x)
        if m > 0.0 and p > 0.0:
            forward = x < p
            camber[forward] = m / p**2 * (2.0 * p * x[forward] - x[forward] ** 2)
            camber[~forward] = (
                m / (1.0 - p) ** 2 * ((1.0 - 2.0 * p) + 2.0 * p * x[~forward] - x[~forward] ** 2)
            )
            camber *= chord
        return camber + thickness, camber - thickness

    def signed_distance(
        self, points: PointCloud, angle_degrees: float
    ) -> Float[np.ndarray, " point"]:
        local = self._to_local(points, angle_degrees)
        upper, lower = self.surfaces(local[:, 0])
        x = local[:, 0]
        y = local[:, 1]
        inside_x = (x >= 0.0) & (x <= self.spec.chord)
        vertical_outside = np.maximum(y - upper, lower - y)
        vertical_inside_distance = -np.minimum(upper - y, y - lower)
        vertical = np.where((y <= upper) & (y >= lower), vertical_inside_distance, vertical_outside)
        dx = np.maximum(-x, x - self.spec.chord)
        outside_x = np.maximum(dx, 0.0)
        distance = np.where(inside_x, vertical, np.hypot(outside_x, np.maximum(vertical, 0.0)))
        return np.asarray(distance, dtype=points.dtype)

    def normals(self, points: PointCloud, angle_degrees: float) -> Float[np.ndarray, "point 2"]:
        epsilon = max(self.spec.chord * 1.0e-4, 1.0e-6)
        offset_x = np.asarray([epsilon, 0.0], dtype=points.dtype)
        offset_y = np.asarray([0.0, epsilon], dtype=points.dtype)
        dx = self.signed_distance(points + offset_x, angle_degrees) - self.signed_distance(
            points - offset_x, angle_degrees
        )
        dy = self.signed_distance(points + offset_y, angle_degrees) - self.signed_distance(
            points - offset_y, angle_degrees
        )
        gradients = np.stack((dx, dy), axis=1)
        lengths = np.linalg.norm(gradients, axis=1, keepdims=True)
        ambiguous = lengths[:, 0] < epsilon
        if np.any(ambiguous):
            angle = np.deg2rad(angle_degrees)
            gradients[ambiguous, 0] = -np.sin(angle)
            gradients[ambiguous, 1] = np.cos(angle)
            lengths[ambiguous, 0] = 1.0
        return gradients / np.maximum(lengths, epsilon)

    def contains(self, points: PointCloud, angle_degrees: float) -> Bool[np.ndarray, " point"]:
        return self.signed_distance(points, angle_degrees) <= 0.0

    def wall_velocity(
        self, points: PointCloud, control: ControlState
    ) -> Float[np.ndarray, "point 2"]:
        relative = points[:, :2] - np.asarray(self.spec.pivot[:2], dtype=points.dtype)
        omega = np.deg2rad(control.angular_velocity_degrees)
        velocity = np.empty_like(relative)
        velocity[:, 0] = -omega * relative[:, 1]
        velocity[:, 1] = omega * relative[:, 0]
        return velocity

    def outline(
        self, angle_degrees: float, samples: int = 256, dtype: np.dtype[np.floating] | None = None
    ) -> Float[np.ndarray, "point 2"]:
        selected_dtype = dtype or np.dtype(np.float32)
        beta = np.linspace(0.0, np.pi, samples // 2, dtype=selected_dtype)
        x = self.spec.chord * 0.5 * (1.0 - np.cos(beta))
        upper, lower = self.surfaces(x)
        local = np.concatenate(
            (
                np.stack((x, upper), axis=1),
                np.stack((x[::-1], lower[::-1]), axis=1),
            ),
            axis=0,
        )
        return self._from_local(local, angle_degrees)

    def mask(self, domain: DomainSpec, angle_degrees: float) -> Bool[np.ndarray, "ny nx"]:
        points = cell_centers(domain)
        return self.contains(points.reshape(-1, 2), angle_degrees).reshape(domain.ny, domain.nx)


def cell_centers(domain: DomainSpec) -> Float[np.ndarray, "ny nx 2"]:
    x0, x1 = domain.bounds[0]
    y0, y1 = domain.bounds[1]
    x = np.linspace(x0 + 0.5 * domain.dx, x1 - 0.5 * domain.dx, domain.nx)
    y = np.linspace(y0 + 0.5 * domain.dy, y1 - 0.5 * domain.dy, domain.ny)
    xx, yy = np.meshgrid(x, y)
    return np.stack((xx, yy), axis=-1)

File: implementations/julia/src/geometry.jl
struct FoilSpec{D,T<:AbstractFloat}
    naca::String
    chord::T
    pivot::SVector{D,T}

    function FoilSpec(
        naca::AbstractString,
        chord::T,
        pivot::SVector{D,T},
    ) where {D,T<:AbstractFloat}
        occursin(r"^[0-9]{4}$", naca) || throw(ArgumentError("NACA code must contain four digits"))
        chord > zero(T) || throw(ArgumentError("foil chord must be positive"))
        return new{D,T}(String(naca), chord, pivot)
    end
end

struct NacaFoil{D,T<:AbstractFloat}
    spec::FoilSpec{D,T}
end

maximum_camber(foil::NacaFoil{D,T}) where {D,T} =
    T(parse(Int, foil.spec.naca[1:1])) / T(100)
camber_position(foil::NacaFoil{D,T}) where {D,T} =
    T(parse(Int, foil.spec.naca[2:2])) / T(10)
thickness(foil::NacaFoil{D,T}) where {D,T} =
    T(parse(Int, foil.spec.naca[3:4])) / T(100)
maximum_radius(foil::NacaFoil{D,T}) where {D,T} = hypot(
    T(0.75) * foil.spec.chord,
    (maximum_camber(foil) + T(0.51) * thickness(foil)) * foil.spec.chord,
)

function surfaces(foil::NacaFoil{D,T}, x_local::AbstractVector{T}) where {D,T}
    chord = foil.spec.chord
    x = clamp.(x_local ./ chord, zero(T), one(T))
    yt = T(5) * thickness(foil) * chord .* (
        T(0.2969) .* sqrt.(max.(x, zero(T))) .-
        T(0.1260) .* x .-
        T(0.3516) .* x .^ 2 .+
        T(0.2843) .* x .^ 3 .-
        T(0.1036) .* x .^ 4
    )
    m = maximum_camber(foil)
    p = camber_position(foil)
    camber = zeros(T, length(x))
    if m > 0 && p > 0
        for index in eachindex(x)
            value = x[index]
            camber[index] = if value < p
                m / p^2 * (T(2) * p * value - value^2)
            else
                m / (one(T) - p)^2 * (
                    (one(T) - T(2) * p) + T(2) * p * value - value^2
                )
            end
        end
        camber .*= chord
    end
    return camber + yt, camber - yt
end

function to_local(foil::NacaFoil{2,T}, point::SVector{2,T}, angle_degrees::Real) where {T}
    angle = T(deg2rad(angle_degrees))
    cosine = cos(angle)
    sine = sin(angle)
    translated = point - foil.spec.pivot
    return SVector(
        cosine * translated[1] + sine * translated[2] + T(0.25) * foil.spec.chord,
        -sine * translated[1] + cosine * translated[2],
    )
end

function signed_distance(foil::NacaFoil{2,T}, point::SVector{2,T}, angle_degrees::Real) where {T}
    local_point = to_local(foil, point, angle_degrees)
    x, y = local_point
    normalized_x = clamp(x / foil.spec.chord, zero(T), one(T))
    thickness_value = T(5) * thickness(foil) * foil.spec.chord * (
        T(0.2969) * sqrt(max(normalized_x, zero(T))) -
        T(0.1260) * normalized_x -
        T(0.3516) * normalized_x^2 +
        T(0.2843) * normalized_x^3 -
        T(0.1036) * normalized_x^4
    )
    camber = zero(T)
    maximum = maximum_camber(foil)
    position = camber_position(foil)
    if maximum > zero(T) && position > zero(T)
        camber = if normalized_x < position
            maximum / position^2 * (T(2) * position * normalized_x - normalized_x^2)
        else
            maximum / (one(T) - position)^2 * (
                one(T) - T(2) * position + T(2) * position * normalized_x - normalized_x^2
            )
        end
        camber *= foil.spec.chord
    end
    upper = camber + thickness_value
    lower = camber - thickness_value
    vertical_outside = max(y - upper, lower - y)
    vertical_inside = -min(upper - y, y - lower)
    vertical = lower <= y <= upper ? vertical_inside : vertical_outside
    if 0 <= x <= foil.spec.chord
        return vertical
    end
    outside_x = max(max(-x, x - foil.spec.chord), zero(T))
    return hypot(outside_x, max(vertical, zero(T)))
end

function signed_distance(
    foil::NacaFoil{2,T},
    points::AbstractMatrix{T},
    angle_degrees::Real,
) where {T}
    size(points, 2) == 2 || throw(DimensionMismatch("points must have shape point × 2"))
    return [signed_distance(foil, SVector{2,T}(points[index, :]), angle_degrees) for index in axes(points, 1)]
end

function normals(foil::NacaFoil{2,T}, points::AbstractMatrix{T}, angle_degrees::Real) where {T}
    size(points, 2) == 2 || throw(DimensionMismatch("points must have shape point × 2"))
    epsilon = max(foil.spec.chord * T(1.0e-4), T(1.0e-6))
    output = Matrix{T}(undef, size(points, 1), 2)
    for index in axes(points, 1)
        point = SVector{2,T}(points[index, :])
        dx = signed_distance(foil, point + SVector(epsilon, zero(T)), angle_degrees) -
             signed_distance(foil, point - SVector(epsilon, zero(T)), angle_degrees)
        dy = signed_distance(foil, point + SVector(zero(T), epsilon), angle_degrees) -
             signed_distance(foil, point - SVector(zero(T), epsilon), angle_degrees)
        gradient = SVector(dx, dy)
        length = sqrt(sum(abs2, gradient))
        if length < epsilon
            angle = T(deg2rad(angle_degrees))
            gradient = SVector(-sin(angle), cos(angle))
            length = one(T)
        end
        output[index, :] = gradient / max(length, epsilon)
    end
    return output
end
